### DISCUSSION TITLE
best practice: prepare data before converting to Data object?

### DISCUSSION THREAD

**Discussion by wivaku** [spatie/laravel-data#423] [Q&A]
What is the best practice to prepare data before creating a nested Data object?
(related to https://github.com/spatie/laravel-data/discussions/414)

I load array data from an external API.
Just as an example, this could be the data I get from that API:
```
$data = [ // collection of PostData
  [
    'title' => 'my post',
    'user' => ['firstName' => 'John', 'lastName' => 'Doe'], // UserData: first + last name
    'timestamp' => 1682168403000, // milliseconds
    'comments' => [ // collection of CommentData
      [
        'title' => 'my comment',
        'user' => ['firstName' => 'Erika', 'lastName' => 'Mustermann'], // UserData: first + last name
        'timestamp' => 1682168147000, // milliseconds
        'replies' => [ // collection of CommentData
          [
          'title' => 'my reply to the comment',
          'user' => ['firstName' => 'John', 'lastName' => 'Doe'], // UserData: first + last name
          'timestamp' => 1682168403000, // milliseconds
          ],
          // more replies...
        ],
        'likes' => [ // collection of UserData
            ['firstName' => 'Erika', 'lastName' => 'Mustermann'], // UserData: first + last name
            ['firstName' => 'Jean', 'lastName' => 'Dupont'], // UserData: first + last name
            // more likes...
        ],
      ],
      // more comments...
    ],
  ],
  // more posts...
]
```

The DTO knows the `PostData` has a `UserData`, and a collection of `CommentData` which consists of `UserData` (and `replies`, which have the same structure as `CommentData`), and `likes` which is a collection of `UserData`. Again, just a (simplified) example.

If the json is in the correct format all I have to do is:
```
$json = loadFromApi($url);
$post = PostData::from($json)
```

But in my case I need to prepare that raw data before creating my object.
E.g. convert durations / timestamps from milliseconds to seconds, or compute fields like name from first+last name.

I prefer that each of the DTO's has the knowledge about preparing that particular DTO (as opposed to preparing the data before handing it over to the `PostData::from(...)`.
E.g. every time a `UserData` object is encountered (in `post`, `comments` or `replies`) get the name from first+last name,
and every time a `CommentData` is encountered (in comment or reply) converts the timestamp from milliseconds to seconds.

But when I create them from a model, then none of these preparations are needed (or even: possible).

What is recommended?

**Comment by Nir-An** [comment]
Hi @wivaku  :)

```
convert durations / timestamps from milliseconds to seconds,
```
you could used [casts](https://spatie.be/docs/laravel-data/v3/as-a-data-transfer-object/casts)


```
name from first+last name
```
you could simply used  the magic method from  and add them together to the same property :+1: 

I hope it is help, I'm encourage you to read the [documentation](https://spatie.be/docs/laravel-data/v3/introduction)
or  see [ a video about it](https://www.youtube.com/watch?v=f4QShF42c6E&t=29110s)
